### PR TITLE
Transformer v2 benchmark

### PR DESCRIPTION
**Pull request by tfboyd**
This includes the refactor from my other PR.  I am not going to merge this until the other PR is merged.

https://github.com/tensorflow/models/pull/6859

### CODE REVIEW

File: official/transformer/v2/transformer_main.py [hunk 1]
```diff
@@ -125,24 +125,38 @@ def train(self):
...
 
     init_epoch = flags_obj.init_epoch or 0
-    init_steps = init_epoch * flags_obj.steps_per_epoch
-    callbacks = self._create_callbacks(cur_log_dir, init_steps, params)
+    init_steps = init_epoch * flags_obj.train_steps
+    callbacks = self._create_callbacks(flags_obj.model_dir, init_steps, params)
 
     history = model.fit(
         train_ds,
-        initial_epoch=init_epoch,
```
Comment (guptapriya) [inline official/transformer/v2/transformer_main.py]: erm, i am not sure if the logic here is all consistent if init_epoch and init_steps are being used.. maybe would be good for @lintian06 to confirm

File: official/transformer/v2/transformer_main.py [hunk 2]
```diff
@@ -125,24 +125,38 @@ def train(self):
...
+        callbacks=callbacks,
+        verbose=2)
     tf.compat.v1.logging.info("\nTrain history: {}".format(history.history))
 
-    save_weight_path = os.path.join(cur_log_dir, "saves-model-weights.hdf5")
-    save_model_path = os.path.join(cur_log_dir, "saves-model.hdf5")
+    stats = misc.build_stats(history, callbacks)
+
+    if flags_obj.bleu_source and flags_obj.bleu_ref:
+      uncased_score, cased_score = evaluate_and_log_bleu(model,
```
Comment (guptapriya) [inline official/transformer/v2/transformer_main.py]: oh, this might be why you saw that error. the model in transformer is different for training vs eval/predict.
you need to create another model here: `model = transformer.create_model(params, is_train=False)`

Although, that makes me think why validation during training works with the same model.. @lintian06 do you know?

File: official/transformer/v2/misc.py [hunk 3]
```diff
@@ -60,113 +65,182 @@ def define_transformer_flags():
...
+
+  Returns:
+    Dictionary of normalized results.
+  """
+  stats = {}
+
+  if history and history.history:
+    train_hist = history.history
+    # Gets final loss from training.
+    stats['loss'] = train_hist['loss'][-1].item()
```
[the remark this answers is not in the document]
Comment (tfboyd) [reply official/transformer/v2/misc.py]: Yeah, I did not see anything that was of interest in history.  The validation is not super useful as we want to bleu score.

File: official/transformer/v2/transformer_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+from official.transformer.v2 import misc
+from official.transformer.v2 import transformer_main as transformer_main
+
+TRANSFORMER_EN2DE_DATA_DIR_NAME = 'wmt32k-en2de-official'
+EN2DE_2014_BLEU_DATA_DIR_NAME = 'newstest2014'
+FLAGS = flags.FLAGS
+
+
+class EstimatorBenchmark(tf.test.Benchmark):
```
[the remark this answers is not in the document]
Comment (tfboyd) [reply official/transformer/v2/transformer_benchmark.py]: Fixing this,  I need to create a base as a big part of this is common, but the reporting part is not.  Maybe I will do that today so we can fix this going forward.

File: official/transformer/v2/transformer_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+                              TRANSFORMER_EN2DE_DATA_DIR_NAME,
+                              'vocab.ende.32768')
+
+    def_flags = {}
+    def_flags['param_set'] = 'base'
+    def_flags['vocab_file'] = vocab_file
+    def_flags['data_dir'] = train_data_dir
+    def_flags['train_steps'] = 200
+    def_flags['train_epochs'] = 1
+    def_flags['epochs_between_evals'] = 10
```
[the remark this answers is not in the document]
Comment (tfboyd) [reply official/transformer/v2/transformer_benchmark.py]: I want to ensure we do not do any evals, I will add a comment.

File: official/transformer/v2/transformer_main.py (hunk 2 above)
Comment (tfboyd) [reply official/transformer/v2/transformer_main.py]: Validation doesn't actually work, well it works but it does not give the values we want.  I guess I don't get why this works just fine with Resnet and we don't set training=False.  I will dig in more today.

File: official/transformer/v2/transformer_main.py (hunk 1 above)
Comment (tfboyd) [reply official/transformer/v2/transformer_main.py]: I would like to see this stream lined to "just work" like the resnet example; we should not need to say what epoch we are starting at to start up training again.  It may take some iterations.

File: official/transformer/v2/transformer_main.py (hunk 2 above)
Comment (guptapriya) [reply official/transformer/v2/transformer_main.py]: if you look inside the `create_model` function, you will see that an different model object is created for training vs eval/predict. it's to do with how custom losses are handled during training - which is not needed in eval so the eval model is slightly simpler. in resnet it's the same model in both cases.

File: official/transformer/v2/transformer_main.py (hunk 1 above)
Comment (tfboyd) [reply official/transformer/v2/transformer_main.py]: I have fixed this and it does not allow starting from a checkpoint.  We can add that back and it should just work and not require you to know what epoch you were on.  The previous code only worked if you knew where you stopped.

File: official/transformer/v2/transformer_main.py (hunk 2 above)
Comment (tfboyd) [reply official/transformer/v2/transformer_main.py]: Fixed

File: official/transformer/v2/transformer_main.py (hunk 1 above)
Comment (lintian06) [reply official/transformer/v2/transformer_main.py]: Hi Toby, the default init_epoch is 0 and just works as another other models, so no need to set or remove.

File: official/transformer/v2/transformer_main.py (hunk 2 above)
Comment (lintian06) [reply official/transformer/v2/transformer_main.py]: Validation is only to give a number. As it also feeds ground-truth labels, it is not a correct way to get BLEU score by bringing in extra knowledge. is_train=False will build another model which is totally agnostic to labels.

File: official/transformer/v2/transformer_main.py [hunk 6]
```diff
@@ -124,31 +123,49 @@ def train(self):
...
-    model.save(save_model_path)
+    callbacks = self._create_callbacks(flags_obj.model_dir, 0, params)
+
+    epochs = 0
+    predict_model = None
+    for i in xrange(1, flags_obj.train_epochs + 1):
+      print("Train Epoch:{}/{}".format(i, flags_obj.train_epochs))
+      history = model.fit(
+          train_ds,
+          initial_epoch=epochs,
```
Comment (guptapriya) [inline official/transformer/v2/transformer_main.py]: this one could be `i-1`?

File: official/transformer/v2/transformer_main.py [hunk 7]
```diff
@@ -124,31 +123,49 @@ def train(self):
...
+          epochs=i,
+          steps_per_epoch=flags_obj.train_steps,
+          callbacks=callbacks,
+          verbose=2)
+      epochs += 1
```
Comment (guptapriya) [inline official/transformer/v2/transformer_main.py]: this can be just `i`?

File: official/transformer/v2/transformer_main.py [hunk 8]
```diff
@@ -124,31 +123,49 @@ def train(self):
...
-        callbacks=callbacks)
-    tf.compat.v1.logging.info("\nTrain history: {}".format(history.history))
-
-    save_weight_path = os.path.join(cur_log_dir, "saves-model-weights.hdf5")
-    save_model_path = os.path.join(cur_log_dir, "saves-model.hdf5")
-    model.save_weights(save_weight_path)
-    model.save(save_model_path)
+    callbacks = self._create_callbacks(flags_obj.model_dir, 0, params)
+
+    epochs = 0
```
Comment (guptapriya) [inline official/transformer/v2/transformer_main.py]: not entirely sure you need another variable. you can write it all in terms of `i` from what i could tell?

File: official/transformer/v2/transformer_main.py [hunk 9]
```diff
@@ -177,23 +194,20 @@ def _create_callbacks(self, cur_log_dir, init_steps, params):
...
+    ckpt_full_path = os.path.join(cur_log_dir, "cp-{epoch:04d}.ckpt")
+    callbacks.append(tf.keras.callbacks.ModelCheckpoint(ckpt_full_path,
+                                                        save_weights_only=True))
+    return callbacks
 
   def _load_weights_if_possible(self, model, init_weight_path=None):
     """Loads model weights when it is provided."""
     if init_weight_path:
       tf.compat.v1.logging.info("Load weights: {}".format(init_weight_path))
-      model.load_weights(init_weight_path, by_name=True)
```
Comment (guptapriya) [inline official/transformer/v2/transformer_main.py]: by_name is not needed anymore?

File: official/transformer/v2/transformer_main.py (hunk 6 above)
Comment (lintian06) [reply official/transformer/v2/transformer_main.py]: Yes. initial_epoch=i-1.

File: official/transformer/v2/transformer_main.py [hunk 10]
```diff
@@ -124,31 +123,49 @@ def train(self):
...
+      epochs += 1
+      tf.compat.v1.logging.info("Train history: {}".format(history.history))
+
+      stats = misc.build_stats(history, callbacks)
+
+      if (flags_obj.bleu_source and
+          flags_obj.bleu_ref and
+          epochs%flags_obj.epochs_between_evals == 0):
+        if not predict_model:
+          predict_model = transformer.create_model(params, False)
```
Comment (lintian06) [inline official/transformer/v2/transformer_main.py]: My concern is that the first model may have name_scope "model", second "model_1", and third "model_2". I don't know whether _load_weights_if_possible can be reloaded properly. Should check.

File: official/transformer/v2/transformer_main.py (hunk 9 above)
Comment (tfboyd) [reply official/transformer/v2/transformer_main.py]: Correct, that was with the other saving style.  I did multiple a 100K step run on 1 GPU and this works correctly.

File: official/transformer/v2/transformer_main.py (hunk 10 above)
Comment (tfboyd) [reply official/transformer/v2/transformer_main.py]: https://sponge.corp.google.com/target?id=48d603e1-252e-45c1-8385-52044f2c5be9&target=tensorflow_models%2Fperfzero%2Fmanual%2Fv100_1_gpu%2Fv100_1_gpu&searchFor=

There is only `model` and `model_1`.  I create the model once and just reload the weights each time.  This saves a little time and looks to work based on the log.

Until I started caching the model, I did some ignored Error in the log but I forget what it was.  But it works fine now.

File: official/transformer/v2/transformer_main.py (hunk 9 above)
Comment (tfboyd) [reply official/transformer/v2/transformer_main.py]: https://sponge.corp.google.com/target?id=48d603e1-252e-45c1-8385-52044f2c5be9&target=tensorflow_models%2Fperfzero%2Fmanual%2Fv100_1_gpu%2Fv100_1_gpu&searchFor=

Yup works perfectly.

File: official/transformer/v2/transformer_main.py (hunk 7 above)
Comment (tfboyd) [reply official/transformer/v2/transformer_main.py]: Sorry about that, much more streamlined now.

File: official/transformer/v2/transformer_main.py (hunk 6 above)
Comment (tfboyd) [reply official/transformer/v2/transformer_main.py]: fixed

File: official/transformer/v2/transformer_main.py [hunk 11]
```diff
@@ -124,31 +123,49 @@ def train(self):
...
-
-    save_weight_path = os.path.join(cur_log_dir, "saves-model-weights.hdf5")
-    save_model_path = os.path.join(cur_log_dir, "saves-model.hdf5")
-    model.save_weights(save_weight_path)
-    model.save(save_model_path)
+    callbacks = self._create_callbacks(flags_obj.model_dir, 0, params)
+
+    epochs = 0
+    predict_model = None
+    for i in xrange(1, flags_obj.train_epochs + 1):
```
[the remark this answers is not in the document]
Comment (tfboyd) [reply official/transformer/v2/transformer_main.py]: I changed it to range.  You likely know the fully story and I also got confused and thought they also had different signatures.  

In Python2 I understood most of the time we should just use xrange; even if the performance gain is minimal.  This creates issues with Python3 so then I started using six.  Bad pattern, with Python2 going away range should be used unless there is a clear performance need due to a huge list.

**Comment by tfboyd** [comment]
@guptapriya  PTAL  please take one more pass.  I think this is much better.  I am sorry I mixed the quote changes and the other refactor into the transformer_main.py change.  Thank you for working through it.  I would rebase; but I know this is close and we can try V2 this afternoon.

**Comment by tfboyd** [comment]
Ok, i'm pushing merge.  I have multiple end-to-end tests and we know there will be more refactors.  Better than it was.  :-)
